### PR TITLE
iOS sound support

### PR DESCRIPTION
**Pull request by ngocdh**
iOS now has stereo sound!

>I had to modify opus lib because of "Compile sources as" "Objective C++" option when compiling for iOS.
-> not anymore thanks to @jeroenvv

Note that for the UI to work nicely (able to quit the settings window for example), #1450 is required.

TODO: Bluetooth headset not working correctly (maybe because of different sample rate). -> or is it, because Android Bluetooth is working fine?

Update: I also added a feature for mobile devices (iOS and Android) in this PR: possibility to switch to built-in mic instead of the external device chosen automatically by the OS, and back. It's helpful when I'm jamming on mobile without external soundcard and want to talk when I'm not playing. More info here: https://github.com/jamulussoftware/jamulus/issues/1512#issuecomment-824713252_

### CODE REVIEW

File: ios/sound.h [hunk 1]
```diff
@@ -29,8 +29,9 @@
 #include "soundbase.h"
 #include "global.h"
 
+#import <Foundation/Foundation.h>
```
Comment (jeroenvv) [inline ios/sound.h:32]: Including Foundation.h and using BOOL made it so I could not compile the project - it causes a lot of 'Expected unqualified-id' errors during parse. 
When taking out this line and changing BOOL to bool in line 60, it works as expected for me.

Comment (ngocdh) [reply ios/sound.h:32]: Oh that's great. Thanks a lot. So you didn't have to "Compile sources as" "Objective C++"? If so we can leave the opus lib untouched.

Is everything working on your ios device?

Comment (ngocdh) [reply ios/sound.h:32]: Yes, now I can compile without setting "Compile sources as" "Objective C++". opus can be left untouched.
And everything still seems to be working fine on my side. Thanks a lot @jeroenvv

Comment (jeroenvv) [reply ios/sound.h:32]: You're welcome :-)

**Comment by jeroenvv** [comment]
I see there is a merge conflict. You can resolve this by following these steps: https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/merging-an-upstream-repository-into-your-fork

Since it is only one file, you could also manually compare the files and reapply your changes to the latest version of this file.

**Comment by ngocdh** [comment]
> I see there is a merge conflict. You can resolve this by following these steps: https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/merging-an-upstream-repository-into-your-fork
> 
> Since it is only one file, you could also manually compare the files and reapply your changes to the latest version of this file.

Thank you. My manual attempt didn't work. I had to merge. Hope it's ok now.

**Comment by softins** [comment]
@ngocdh Thank you for this work! I'd like to review it. I have compiled it on Linux and verified that it still works as expected.

How do I compile it on my Mac using xcode and then test it on my iPad?

**Comment by ngocdh** [comment]
> @ngocdh Thank you for this work! I'd like to review it. I have compiled it on Linux and verified that it still works as expected.
> 
> How do I compile it on my Mac using xcode and then test it on my iPad?

Sure. I’m pasting the response from @jeroenvv which helped me:

>What I found out was that I needed to install qt5 with the qt installer (not homebrew), and explicitly select iOS when choosing the Qt version.
>After installation, I ran /5.12.10/ios/bin/qmake -spec macx-xcode Jamulus.pro
>Only with the right qmake, I got all the necessary libraries correctly configured. Then, in XCode, check the Signing & Capabilities and configure a team if you want to run on a device.
>After that, I could run from XCode.

**Comment by ngocdh** [comment]
Copy and run on ipad is straightforward. You just have to connect your ipad, choose ipad as target when Run. And on your ipad, go to settings and trust your developer profile.
Ah and if you haven’t already, add your apple account in Xcode so it can sign the app.

**Comment by ngocdh** [comment]
@softins were you able to test on your iPad?

**Comment by softins** [comment]
> @softins were you able to test on your iPad?

Sorry, life is a bit crazy at the moment due to family circumstances, and I haven't had time to try yet.

**Comment by ngocdh** [comment]
> > @softins were you able to test on your iPad?
> 
> Sorry, life is a bit crazy at the moment due to family circumstances, and I haven't had time to try yet.

Oh I hope things sort out for you @softins

**Comment by ngocdh** [comment]
Team, I've been implementing a new feature (for mobile devices): possibility to switch to built-in mic instead of the external device (which is chosen automatically by the OS), and back. It's helpful when I'm jamming on mobile (without external soundcard) and want to talk when I'm not playing.

This feature is already implemented in this pull request for iOS.

For Android though, it's a bit more complicated. I managed to make it work, but the app needs help from the user to input the correct deviceid of the built-in mic. This deviceid is not the same for different devices and oboe has no way to obtain it. Java must be involved. So for now to avoid the complication, I've added a textbox so that the user can input the deviceid (which is a integer number between 0 and 100), which they can obtain from an external java audio manager app. 

So for Android it's still experimental, but working. I implemented this in a branch, but the branch also includes modifications made from this pull request, because I had to make sure the feature worked on both iOS and Android.

My question is, do I commit here directly, or do I create a new pull request, knowing that new pull request will include modifications from this one, which is not yet approved?

**Comment by softins** [comment]
> My question is, do I commit here directly, or do I create a new pull request, knowing that new pull request will include modifications from this one, which is not yet approved?

I would say that as it is all part of the same project of enhancing iOS sound support, if you have more to add, mark the PR as Draft (to prevent premature merging) and push to the same branch. It can still be reviewed and tested by others while Draft, then when you have got it to the state you want, you can mark it as Ready again.

**Comment by ngocdh** [comment]
> I would say that as it is all part of the same project of enhancing iOS sound support, if you have more to add, mark the PR as Draft (to prevent premature merging) and push to the same branch. It can still be reviewed and tested by others while Draft, then when you have got it to the state you want, you can mark it as Ready again.

Thanks @softins for the answer. I think I just made a mistake when trying to follow your instructions: I first converted my PR in the dev branch to Draft, but then couldn't figure out how to push the Draft PR to my main branch. So I pushed the dev branch to my main branch hoping the Draft part would stay Draft. 

Now the Draft PR is still there, but my main branch is updated.

Anyway, if this PR gets approved we should be fine.

**Comment by softins** [comment]
@ngocdh ah, I think the mistake was in the first place doing all the development in your `master` branch and submitting the PR from there. The correct technique for developing a PR is:
1. Make sure your `master` branch is up to date with the upstream `master`, and do `git checkout master` to make sure it is the current branch.
2. Create a new branch for your changes, by doing `git checkout -b my-new-feature`.
3. Develop and commit your changes on the `my-new-feature` branch.
4. Push that branch to your `origin` repo.
5. Go to github.com and create a PR from your `my-new-feature` branch.
6. If you ticked "maintainers can push changes", then the main developers can if necessary push a change to your `my-new-feature` branch without disturbing your `master`.
7. Once your feature has been merged into the upstream `master`, you can get your own `master` up to date, and then delete your `my-new-feature` branch.

Also, I'm not sure exactly how you "converted your dev branch to Draft". It was this PR on github that needed converting to Draft. Look at the top right under "Reviewers", where it says "Still in progress? Convert to Draft" and click on that.

**Comment by ngocdh** [comment]
@softins thanks a lot for the explanation. I'll keep that in mind.

> Also, I'm not sure exactly how you "converted your dev branch to Draft". It was this PR on github that needed converting to Draft. Look at the top right under "Reviewers", where it says "Still in progress? Convert to Draft" and click on that.

Yeah, I did try to find that "Still in progress? Convert to Draft" in this PR, but didn't find it (not available for non members?). The only place I found that was in the PR where I pushed changes from the dev branch to my main branch.

**Comment by softins** [comment]
Ah, I see what you did now. You created a PR from your mobile mic branch to _your own_ master branch, rather than to the upstream repo. It was that PR that you marked as Draft.

I can mark this PR as Draft myself until we can understand the confusion and get it sorted out!

**Comment by ngocdh** [comment]
> Ah, I see what you did now. You created a PR from your mobile mic branch to _your own_ master branch, rather than to the upstream repo. It was that PR that you marked as Draft.

Correct

> I can mark this PR as Draft myself until we can understand the confusion and get it sorted out!

Thanks. This PR contains everything I've developed, including iOS sound, the built-in mic feature for iOS, and then Android. What do we need to mark it as "Ready for review"?

**Comment by ngocdh** [comment]
> Not sure yet. I'm going to fetch from your own fork and do some checks. By the way, I just merged #1450 into master.

Great!

**Comment by nefarius2001** [comment]
Hey,
I'm really excited to see IOS development here. And eager to try it out myself :). But how do I build and try it? I have no mac dev machine, so I would like to set up some autobuild for that (github or docker). 

Is there already some kind of build script? If not, my guess for the procedure is:
1) use mac container
2a) install xcode and qt  << for developers
2b) use aqt to install qt for ios << for containers
3) execute qmake 
==> any additional parameters?
==> do i get an IPA or DEB file directly?

**Comment by ngocdh** [comment]
> Hey,
Hey @nefarius2001,

> I'm really excited to see IOS development here. And eager to try it out myself :). But how do I build and try it? I have no mac dev machine, so I would like to set up some autobuild for that (github or docker). 

Great to hear that. I've been using Jamulus on iPhone to listen to jams, and even play sometimes. A very good 5Ghz wifi router is required though to overcome the delay issue. Every week I have to rebuild the app, because of the expired signature, as I don't have Apple dev account.

Having no experience in iOS distribution, I have no idea how to release the iOS app. It seems that for testing, a developer must enroll in the paid program to sign an app and send it to a very limited number of testers.
 

> Is there already some kind of build script? If not, my guess for the procedure is:
> 
> 1) use mac container
> 
> 2a) install xcode and qt  << for developers
> 
> 2b) use aqt to install qt for ios << for containers
> 
> 3) execute qmake 
> 
> ==> any additional parameters?
> 
> ==> do i get an IPA or DEB file directly?
> 

Even if autobuild is possible (meaning hardcoded signing?), we still have that limit on the number of devices.

AppStore and GPL don't seem to like each other...

Again I have really no experience here, so would love to hear opinions, and maybe a solution?

**Comment by nefarius2001** [comment]
Hi,
i have three goals:

1.  get it running on my phone for curiosity
2. make it accessible to "beta-users", no matter what the state of development is and no matter if there are extra steps to install (like the one week thing )
3. make it easily accessible to standard users (likely the store, but lets start that discussion once there is some tested version ready to distribute)

But let me start with 1): Can you (@ngocdh) describe in a few sentences with what tools/IDEs you build and how you install on the device?

**Comment by ngocdh** [comment]
> But let me start with 1): Can you (@ngocdh) describe in a few sentences with what tools/IDEs you build and how you install on the device?

You'll need a Mac, Xcode and Qt5 for iOS. See https://github.com/jamulussoftware/jamulus/pull/1512#issuecomment-821314799
Qt5 is needed to generate Xcode project. To build and run on your iOS devices, Xcode need to sign the app with your apple id.
Then in your device, go to Settings and trust the dev profile. And that's it.

**Comment by nefarius2001** [comment]
Ok, thanks

**Comment by ngocdh** [comment]
Just curious, has anyone successfully built and ran this pull request?

**Comment by softins** [comment]
It’s still on my todo list, bu sadly hasn’t reached the top yet!

**Comment by ann0see** [comment]
> Having no experience in iOS distribution, I have no idea how to release the iOS app. It seems that for testing, a developer must enroll in the paid program to sign an app and send it to a very limited number of testers.

Yes. Unfortunately, you must pay to distribute it via the AppStore. GPL and Apple don't work well, Nextcloud has found a way 
around it: https://nextcloud.com/blog/nextcloud-ios-app-open-sourced/ I doubt Jamulus would be accepted at this stage because of the missing mobile UI. 

There are multiple ways to install Apps not part of the AppStore. However, they are not something a typical user would like to go through. 

1. Sideloading with AltStore, CydiaImpactor, ... they all need an ipa file which you'd sign on your PC.
2. Jailbreaking (at least for beta users). You'd create a .deb file which you install via Filza File manager or dpkg via Terminal. But that's nothing an average user would like to do.

**Comment by ngocdh** [comment]
> GPL and Apple don't work well

I just found a ton of GPL apps on the app store. Google dork: `site:apps.apple.com gplv3` or even `site:apps.apple.com gplv2`. So maybe it's not that incompatible after all? 

>I doubt Jamulus would be accepted at this stage because of the missing mobile UI.
 
You're talking about Apple approval process?

> There are multiple ways to install Apps not part of the AppStore. However, they are not something a typical user would like to go through.

Yeah only advanced users for now.

> Edit: Don't have enough space on my SSD to update Xcode...

Too bad...

Now I'm not sure yet, but let's say we can publish on the App Store. It seems the only requirement is a 99$ developer account? Anyone with such account can build Jamulus and publish, or do we really need to buy an account on behalf of the project?

**Comment by ngocdh** [comment]
So it seems (for beta testing at least), any dev account should be able to build and upload to TestFlight, no approval process involved.
https://developer.apple.com/support/app-store-connect/

It would be helpful if anyone with a dev account can help with this. Not sure though if the team is ok?

**Comment by nefarius2001** [comment]
> So it seems (for beta testing at least), any dev account should be able to build and upload to TestFlight, no approval process involved.
> https://developer.apple.com/support/app-store-connect/
> 
> It would be helpful if anyone with a dev account can help with this. Not sure though if the team is ok?

Are there anyone with an apple dev account voluteering already, or still out there to raise their hand? (sorry if I missed it)

**Comment by ngocdh** [comment]
Hi @nefarius2001 I'm hoping someone with an Apple dev account will volunteer.

Update about beta testing: it's limited to 10 team members per account, no review required. 10000 "external testers", but the app must go through review process first, app expired after 30 days.

**Comment by nefarius2001** [comment]
Hi,
does any of the active developers have an Apple Dev account or know someone with such?
I'd love iOS support to have the opportunity for mobile setups. having apple AND android covered would do that job.

(Personally still struggling to get hold of a mac machine for compiling...)

**Comment by ann0see** [comment]
> Personally still struggling to get hold of a mac machine for compiling...

You could try a macOS VM there’s a Script somewhere for macOS on QEMU/KVM for Linux

**Comment by hoffie** [comment]
>  does any of the active developers have an Apple Dev account or know someone with such?

I don't.

I think going forward with further proof of concept work which is usable using non-official installation methods would still be beneficial. There will be a risk that the app might never become official, but it may still help some users.

**Comment by ngocdh** [comment]
> I think going forward with further proof of concept work which is usable using non-official installation methods would still be beneficial. There will be a risk that the app might never become official, but it may still help some users.

In case no active developer has an Apple dev account, if we have someone volunteering in publishing Jamulus to the App Store (not that I have anyone in mind yet), would the team be ok? Know that the App Store could mean a few complications to comply with Apple's requirements, and maybe the app will be published under their name, not "jamulussoftware".
 
Anyway, could there be at least a way to simplify the developing of the iOS app? For now, one must have a Mac, install Xcode, Qt5, generate the xcodeproject to be able to open the project in Xcode...

> Note: This PR currently has conflicts. There may be two main reasons:
> 
> * The PR contains .qm updates. This is unnecessary. They can be removed (`git reset`) from this PR.
> * Right before 3.8.0rc1, we've merged a PR which performs consistent code formatting on the whole code base. Running this formatter (`clang-format`) on this PR should also improve the chances for conflicts to go away without too much manual work.


Could you show me exactly what to do to perform both actions?

**Comment by alexbuckland** [comment]
We would like to donate someone the developer account fee if it helps move this forward.

**Comment by ann0see** [comment]
Ok. Compiled it and it crashed because of a missing LaunchScreen file. After I opened the app once, it worked. Afterwards it crashed again.

**Comment by ngocdh** [comment]
> Ok. Compiled it and it crashed because of a missing LaunchScreen file. After I opened the app once, it worked. Afterwards it crashed again. 

Great I'm glad you managed to test it. On my iPhone, the app does crash from time to time, when I click on "Connect". When you say "it worked", you mean the audio and everything else too? What led to the 2nd crash?

**Comment by ngocdh** [comment]
@jeroenvv could you help with the LaunchScreen crash?

**Comment by ann0see** [comment]
Yes. Audio worked. I can only open the app once after compiling.

I made some audio tests on my iPhone SE 1st gen with Interface & LAN and audio quality was good. Maybe some more dropouts occurred, but that’s acceptable. 

793FD921-C6A3-4410-8380-169C108264BD

Latency might be a bit higher than on Linux Debian Testing, but probably not much

**Comment by ngocdh** [comment]
> Yes. Audio worked. I can only open the app once after compiling.
> 
> I made some audio tests on my iPhone SE 1st gen with Interface & LAN and audio quality was good. Maybe some more dropouts occurred, but that’s acceptable.
> 
> 793FD921-C6A3-4410-8380-169C108264BD
> 
> Latency might be a bit higher than on Linux Debian Testing (~3ms), but probably not much

Awesome, those are great numbers for a 2016 iphone! I usually run Jamulus on an iphone 12, good 5Ghz wifi, and without opus (don't see much difference though).

Notice you have a feature not present on "normal" Jamulus: the "Mobile input" menu, which allows you to change between the built-in mic and external input device (for instrument).

Now we have to work those crashes out, which for me mostly happen upon loading and clicking on "Connect". iOS-related UI bugs?

**Comment by ann0see** [comment]
>  iOS-related UI bugs?

Maybe. Which Qt version do you use? I used the latest Qt5 version available (5.15.2). I didn't have any crashes on connect. Only on the second, third,... launch which makes the app unusable. For me it says "Failed to terminate gracefully after 10.0s or exhausted real (wall clock) time allowance of 19.61 seconds in the iOS crash logs"

**Comment by ngocdh** [comment]
My version is 5.12.10. Where can I find the iOS crash logs?

**Comment by ngocdh** [comment]
> Will try this Qt version too. 

Don't, 5.15.2 is working great for me. No crash. Just read my last comment.

> You can find the crash logs in iOS settings under privacy. 

Thanks.

**Comment by ann0see** [comment]
Ah. Missed it. Will try again

**Comment by ann0see** [comment]
There’s also another way: check the macOS console while your phone is connected

**Comment by ann0see** [comment]
Ok. Maybe the app just doesn't launch fast enough: https://stackoverflow.com/questions/4712333/iphone-how-can-i-avoid-the-watchdog-killing-my-application-for-taking-long-time
iOS seems to kill slow apps.

**Comment by ngocdh** [comment]
Makes sense! Maybe we need to optimize and remove all the unnecessary initializations. On Windows, Jamulus also takes long to load.

**Comment by ngocdh** [comment]
> On Windows, Jamulus also takes long to load.

I just verified this statement. Turns out Asio4all loads fast, but if I choose my other device, "Yamaha Steinberg USB ASIO", it takes a lot longer.

So I did a simple test: check the load time difference in iOS with/without sound (2 different builds). Result: while load time without sound can't be considered good, it's a lot faster than with sound.

Sound initialization might be the problem here.

**Comment by ngocdh** [comment]
@ann0see could you try the new commit? Launch time should be a lot better now.

**Comment by ann0see** [comment]
Puh. That’s a lot. Can you open an issue with your findings?

**Comment by ngocdh** [comment]
Here you are #1727 @ann0see.

**Comment by ann0see** [comment]
Great! Thanks

**Comment by ngocdh** [comment]
So "Connect" still crashes, and I managed to reproduce this behavior: open Jamulus, leave it for a while, maybe open other apps in the meantime. Then return to Jamulus, click on Connect -> crash.
Looks like a use-after-free issue.

**Comment by ann0see** [comment]
Hmm. So nothing changed (the last commit seems to revert your changes)? Can the checks be done after the App launched?

**Comment by ngocdh** [comment]
> Hmm. So nothing changed (the last commit seems to revert your changes)? 

I reverted half of it, meaning it's now 9 inits, instead of 36. So you'll already get a better launch time.

> Can the checks be done after the App launched?

Yes I believe, but maybe a deep analysis is required to find the best spot in the code to change this.

**Comment by ann0see** [comment]
… multithreading …

Probably.

 I tested this code a few days ago and the app still crashed.

**Comment by ngocdh** [comment]
@ann0see you can try the 1 init build https://github.com/jamulussoftware/jamulus/pull/1512/commits/3cbb2c404a354aae3d1d698b8cc287ad8b403ff5 to see if it launches. If it does, then you’ll only be able to use buffer size 128 (which is why I changed back to 9 inits). Still, if it works then you’ll know if sound inits should be optimized to welcome more devices.

**Comment by ann0see** [comment]
Yep. Although the app still takes a long time to start, it starts with on this commit for me.

**Comment by ann0see** [comment]
Maintenance stuff: Related: https://github.com/jamulussoftware/jamulus/discussions/1259

**Comment by ngocdh** [comment]
> Yep. Although the app still takes a long time to start, it starts with on this commit for me.

Great, so sound inits optimization we need.

> The sound quality of the internal iPhone mic is quite impressive compared to my PC and headphones.

Yes.

**Comment by ngocdh** [comment]
Update: I removed some unused libs, and the app seems to load a bit faster for me.

**Comment by ann0see** [comment]
Yes. This seems to fix the crash. So no MIDI support for iOS probably?

**Comment by ann0see** [comment]
Ok. This PR might be ready for the next release (not 3.8.0)

I‘ll leave this PR alone for now and would be happy if someone else took over. Will need to focus on the website.

What should be checked:

- [ ] Code style (see clang format); remove commented code
- [ ] Basic code review: what can still be improved? Is the logic ok, can improvements be made?
- [ ] Improve the commit history of this PR (or squash & merge it later on?)
- [ ] Decide about code signing and distribution
- [ ] What should we do about UI issues

**Comment by ngocdh** [comment]
Been trying to fix the Connect button crash without success. Turns out it has to do with iOS sockets handling when device goes into idle mode.

Ref: https://www.b4x.com/android/forum/threads/udpsocket-silent-crash-of-app-after-using-iphone-standby-button.85565/

File: src/clientdlg.cpp [hunk 2]
```diff
@@ -1093,6 +1106,12 @@ void CClientDlg::OnTimerCheckAudioDeviceOk()
                                tr ( "Your sound card is not working correctly. "
                                     "Please open the settings dialog and check the device selection and the driver settings." ) );
     }
+    /*else
+    {
+      #if defined ( ANDROID ) || defined ( Q_OS_ANDROID )
+        QMessageBox::warning ( this, APP_NAME, pClient->Sound.getAvailableDevices() ); //DEBUG NGOCDH
+      #endif
+    }*/
```
Comment (ann0see) [inline src/clientdlg.cpp:1114]: Still needed?

File: src/socket.cpp [hunk 3]
```diff
@@ -45,6 +50,12 @@ void CSocket::Init ( const quint16 iPortNumber, const quint16 iQosNumber, const
     const char tos = (char) iQosNumber; // Quality of Service
     setsockopt ( UdpSocket, IPPROTO_IP, IP_TOS, &tos, sizeof ( tos ) );
 
+#ifdef Q_OS_IOS
+    //ignore the broken pipe signal to avoid crash (iOS)
+    int valueone = 1;
+    setsockopt(UdpSocket, SOL_SOCKET, SO_NOSIGPIPE, &valueone, sizeof(valueone));
```
Comment (ann0see) [inline src/socket.cpp:56]: Styling

File: src/socket.cpp [hunk 4]
```diff
@@ -188,12 +202,24 @@ void CSocket::SendPacket ( const CVector<uint8_t>& vecbySendBuf, const CHostAddr
...
-                 (sockaddr*) &UdpSocketOutAddr,
-                 sizeof ( sockaddr_in ) );
+        if (
+            sendto ( UdpSocket,
+                     (const char*) &( (CVector<uint8_t>) vecbySendBuf )[0],
+                     iVecSizeOut,
+                     0,
+                     (sockaddr*) &UdpSocketOutAddr,
+                     sizeof ( sockaddr_in ) )
+            < 0 )
```
Comment (ann0see) [inline src/socket.cpp:212]: style

**Review by ann0see — CHANGES_REQUESTED** [review]

File: src/clientdlg.cpp (hunk 2 above)
Comment (ngocdh) [reply src/clientdlg.cpp:1114]: Seems like a test at an earlier stage of the dev that I don't even remember lol. Will remove it.

File: src/socket.cpp (hunk 4 above)
Comment (ngocdh) [reply src/socket.cpp:212]: How do I run the Coding Style check locally?

Comment (ann0see) [reply src/socket.cpp:212]: You can do it with clang-format. I‘m not sure about the exact command, but maybe it’s something with clang-format -e ? However, you‘ll probably need to update this PR to the current master (= rebase your git branch).

Comment (ngocdh) [reply src/socket.cpp:212]: Ok I’ll check clang-format, and also what rebase means lol

Comment (ann0see) [reply src/socket.cpp:212]: Maybe this helps: https://learnxinyminutes.com/docs/git/

**Comment by ann0see** [comment]
Hmm. Something seems to be wrong with this PR (the diff seems to look wrong). You should maybe apply your code on a clean branch again (and generate separate commits) afterwards open a new PR or git push --force it to your master branch. Have a look at the translation file: https://github.com/jamulussoftware/jamulus/blob/master/TRANSLATING.md which might help you getting the commit history right.

**Comment by ngocdh** [comment]
Ok I'll try that.

**Comment by ann0see** [comment]
I've saved your changes now here: https://github.com/ann0see/jamulus/tree/ngSave

**Comment by ngocdh** [comment]
Thanks! I'm thinking about creating a brand new branch from jamulus/master, then merge from a blob from my branch before today's commits, without those language files. Do you think it's possible?

**Comment by ann0see** [comment]
No. You should not merge your changes manually. I assume this will make it worse. The cleanest (and easiest way) would probably be to get your code and apply it manually on a new branch. 

It's probably possible to fix it somehow, but I'm not that confident with git to tell you exactly how to fix this problem.

I know @hoffie is quite good at sorting git out. Maybe he can help you?

The best way is not a merge, but a rebase if you want to update your branch. It's more complicated sure. Since I also don't often use git rebase, I would probably get me a new branch, cherry pick the commits I want and then force push this branch. However I doubt this is a good solution here.

**Comment by ngocdh** [comment]
> I would probably get me a new branch, cherry pick the commits I want and then force push this branch.

That's kind of like what I meant. Will try it first.

**Comment by ann0see** [comment]
I think something like git rebase -i might work (search the net for interactive rebase).

**Comment by pljones** [comment]
This still seems like a completely invalid commit, given what it's _meant_ to be doing.

```
$ git status
```
That should show your current status on your current branch.  When you're ready, you should have all your work committed to your branch locally, so there should be no changed files locally.

Ideally then do a `git diff` against the base commit of your branch - i.e. where you branched from the trunk of master - to check the changes are what you expect and only what you expect.  It may be worth getting your local master back to that point and doing a rebase with squash, to reduce the number of commits, too.

Once you're happy with what you have on your branch...
```
$ git checkout master
$ git pull upstream master; git pull upstream master --tags --force
$ git push; git push --tags --force
```
This brings your _local_ repository up to date with `jamulussoftware/jamulus` (assuming you've used `upstream` as the remote name for that) and the pushes those changes to your github repository.
```
$ git checkout -
```
...gets you back to your branch.
```
$ git rebase -i master
```
The reason for the `-i` is to get the list of commits that git think need applying.  They should _only_ be the commits you've just checked above.  Anything else, delete it.  Note that this _can_ go horribly wrong, so it can be worth taking a copy of the whole of your working directory...

The rebase can also, quite legitimately, show conflicts.  However, I'd be surprised here, as this should be fairly isolated code, right?

**Comment by ngocdh** [comment]
I’ve got a new clean working branch here https://github.com/ngocdh/jamulus/tree/ngocdhmaster2. Should I just create a new pull request and ditch this one? Simpler, cleaner?

**Comment by gilgongo** [comment]
Closing and replacing with #1865

**Pull request closed without merging**